core/system/src/context.c

```diff
@@ -0,0 +1,266 @@
+    return dst_sp;
+}
+
+/* Discards an unused exception stack frame from the destination box. */
+void context_discard_exc_sf(uint8_t dst_id, uint32_t dst_sp)
+{
+    uint32_t exc_sf_alignment;
+
+    exc_sf_alignment = (((uint32_t *) dst_sp)[7] & 0x4) ? 1 : 0;
+    g_context_current_states[dst_id].sp = dst_sp + CONTEXT_SWITCH_EXC_SF_SIZE + exc_sf_alignment;
+}
+
+/** Validates an exception stack frame, checking that the currently active box
+ *  has access to it.
+ *
+ * @internal
+ *
+ * The exception stack frame is validated through direct access: If the stack
+ * frame starting from the provided pointer is not fully accessible in
+ * unprivileged mode, it means that the currently active box does not fully own
+ * that memory section. This results in an access fault. The check covers the
+ * following addresses:
+ *  pointer[0] to pointer[7] if the stack frame is aligned
+ *  pointer[0] to pointer[8] if the stack is not aligned */
+uint32_t context_validate_exc_sf(uint32_t exc_sp)
+{
+    int i;
+
+    /* Check the regular exception stack frame. */
+    for (i = 0; i < CONTEXT_SWITCH_EXC_SF_SIZE; i++) {
+        vmpu_unpriv_uint32_read(exc_sp + sizeof(uint32_t) * i);
+    }
+
+    /* Read one additional word if the stack frame is not aligned. */
+    if (!(exc_sp & 0x4)) {
+        vmpu_unpriv_uint32_read(exc_sp + sizeof(uint32_t) * CONTEXT_SWITCH_EXC_SF_SIZE);
+    }
+
+    return exc_sp;
+}
+
+/* Switches the context from the source box to the destination one, using the
+ * stack pointers provided as input. */
+void context_switch_in(TContextSwitchType context_type, uint8_t dst_id, uint32_t src_sp, uint32_t dst_sp)
+{
+    uint8_t src_id;
+
+    /* Source box: Gather information from the current state. */
+    src_id = g_active_box;
+
+    /* The source/destination box IDs can be the same (for example, in IRQs). */
+    if (src_id != dst_id) {
+        /* Update the context pointer to the one of the destination box. */
+        *(__uvisor_config.uvisor_box_context) = (uint32_t *) g_context_current_states[dst_id].context;
+
+        /* Update the ID of the currently active box. */
+        g_active_box = dst_id;
+
+        /* Switch MPU configurations. */
+        /* This function halts if it finds an error. */
+        vmpu_switch(src_id, dst_id);
+    }
+
+    /* Push the state of the source box and set the stack pointer for the
+     * destination box.
+     * This is only needed if the context switch is tied to a function. Unbound
+     * context switches require the host OS to set the correct stack pointer
+     * before handling execution to the unprivileged code, and for the same
+     * reason do not require state-keeping.  */
+    /* This function halts if it finds an error. */
+    if (context_type == CONTEXT_SWITCH_FUNCTION_GATEWAY ||
+        context_type == CONTEXT_SWITCH_FUNCTION_ISR     ||
+        context_type == CONTEXT_SWITCH_FUNCTION_DEBUG) {
+        context_state_push(context_type, src_id, src_sp);
```

Comment:
Just an aside on our style: I would find these sorts of functions more readable with the `{` on a new line below the `if` predicates.